### ISSUE TITLE
An easy way to toggle debugger() calls ?

### ISSUE THREAD

**Issue by sethladd** [dart-lang/sdk#23939]
Hi,

I'm using Observatory and the cool debugger() call. When debugging, it's quite helpful. However, when I'm switching between "just run to see what happens" and "debug", I need to comment out all of my debugger calls. That's possible, but wondering if we can make it a bit easier...

If there was a way to opt-in to respecting the debugger() calls, I could do this:

`dart --enable-debugger --observe main.dart`

or

`dart main.dart`

In the first case, I want to observe and trigger debugger() calls. In the second case, I just want my app to run.

This is just a strawman idea... but would love a way to turn on or off the debugger() calls without having to comment them all on and off.

Thanks!

**Comment by sethladd** [comment]
cc @johnmccutchan for ideas

**Comment by johnmccutchan** [comment]
What worries me is that you'll forget about the debugger() calls if we have the flag.

Note that debugger() takes a 'when' parameter: debugger(when: globalDebugFlag);

**Comment by sethladd** [comment]
> What worries me is that you'll forget about the debugger() calls if we have the flag.

True. But since a developer has to actively tell their VM to turn those on, what happens if I forget about them?

> Note that debugger() takes a 'when' parameter: debugger(when: globalDebugFlag);

Yeah, which is great. Except I use that to break conditionally on some state. :)

We can chat on Monday... maybe there's a workaround. For now, not blocked.

**Comment by lrhn** [comment]
It seems like a feature request for a VM flag rather than a library issue.

Another option, which we should consider, is to not hang if there is no debugger attached.
The JS `debugger` expression continues running if there is no debugger attached, and I think that would be a reasonable behavior. If possible the VM should just do the same. 

@sethladd You could make a helper library for this:
```dart
library debuggerhelper;
import "dart:developer" as dev;

get debugger => const String.fromEnvironment('DEBUG') == null ? _nodebugger : dev.debugger;
_nodebugger({when,message}) {}
```

**Comment by sethladd** [comment]
> You could make a helper library for this

Thanks, that helps. I'm imagining that most developers who end up using `debugger()` will end up wanting this. I'd like to ask that we put this into `dart:developer`.

**Comment by lrhn** [comment]
I think it would be better to just change the VM hehavior when calling `debugger` with no debugger attached. Then you don't need any extra functions or flags at all.

**Comment by johnmccutchan** [comment]
@lrhn The VM always has a debugger attached. Also, we rely on debugger() working before someone connects over the service protocol.

**Comment by lrhn** [comment]
Ok, I can see there is no good way to ignore the debugger function.
How about a flag to disable the debugger function. It should be off by default, but you can pass it while developing to run a program without debugger() interrupting - so you don't have to remove the debugger callas until you know the program is working.

I don't think we should use a -Ddisable-debugger=true flag, I'd prefer it to be a real VM flag, and keep the -D flags open for users.
